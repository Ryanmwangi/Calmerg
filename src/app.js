import server from './server.js';

const port = process.env.NODE_PORT || 3012;
server.listen(port, () => {
    console.log(`Server started on port ${port}`);
});
