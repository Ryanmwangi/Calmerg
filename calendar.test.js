import request from 'supertest';
import express from 'express';
import fs from 'fs';
import path from 'path';
import app from './server';

// Set environment variable for the test directory
process.env.TEST_MERGED_CALENDARS_DIR = path.join(__dirname, 'temp_test_calendar');

const TEST_MERGED_CALENDARS_DIR = process.env.TEST_MERGED_CALENDARS_DIR;
const TEST_CALENDARS_DIR = 'test_calendars';
const EXPECTED_OUTPUTS_DIR = 'expected_outputs';
console.log(`Test Merged Calendars Directory: ${TEST_MERGED_CALENDARS_DIR}`);
let server;

describe('Calendar Merging API', () => {
    beforeAll(async () => {
        // Change the working directory to the test-specific directory
        process.chdir(path.join(__dirname, 'temp_test_calendar'));
        // Start the server
        server = app.listen(0);
        // Ensure the test merged calendars directory exists
        if (!fs.existsSync(TEST_MERGED_CALENDARS_DIR)) {
            fs.mkdirSync(TEST_MERGED_CALENDARS_DIR, { recursive: true });
        }
    });

    afterAll(async () => {
        // // Clean up the merged calendars directory after tests
        // if (fs.existsSync(TEST_MERGED_CALENDARS_DIR)) {
        //     fs.rmSync(TEST_MERGED_CALENDARS_DIR, { recursive: true, force: true });
        // }
        
         // Close the server
        await new Promise(resolve => server.close(resolve));
    });

    const loadCalendarFile = (filename) => {
        return path.join(__dirname, TEST_CALENDARS_DIR, filename);
    };

    const loadExpectedOutput = (filename) => {
        return fs.readFileSync(path.join(EXPECTED_OUTPUTS_DIR, filename), 'utf8');
    };

    test('Merge date-based calendar', async () => {
        const response = await request(server)
            .post('/merge')
            .send({
                linkGroupName: 'Date Based Calendar',
                calendars: [
                    {
                        url: loadCalendarFile('ferien_bayern_2023.ics'),
                        prefix: 'Ferien_Bayern_2023',
                        override: false,
                    },
                    {
                        url: loadCalendarFile('US_Holidays.ics'),
                        prefix: 'US_holidays',
                        override: false,
                    },
                ],
            });
        expect(response.status).toBe(200);
        expect(response.body.url).toMatch(/calendar\/Date_Based_Calendar/);

        // Check if the file was created in the test directory
        const filePath = path.join(TEST_MERGED_CALENDARS_DIR, 'Date_Based_Calendar.ics');
        expect(fs.existsSync(filePath)).toBe(true);

        // Load expected output and compare
        const expectedOutput = loadExpectedOutput('Date_Based_Calendar.ics');
        const actualOutput = fs.readFileSync(filePath, 'utf8');
        // expect(actualOutput).toBe(expectedOutput);
    });

    test('Merge time-based calendar', async () => {
        const response = await request(server)
            .post('/merge')
            .send({
                linkGroupName: 'Time Based Calendar',
                calendars: [
                    {
                        url: loadCalendarFile('other_work.ics'),
                        prefix: 'other_work',
                        override: false,
                    },
                    {
                        url: loadCalendarFile('work.ics'),
                        prefix: 'work',
                        override: false,
                    },
                ],
            });

        expect(response.status).toBe(200);
        expect(response.body.url).toMatch(/calendar\/Time_Based_Calendar/);
        
        // Check if the file was created in the test directory
        const filePath = path.join(TEST_MERGED_CALENDARS_DIR, 'Time_Based_Calendar.ics');
        expect(fs.existsSync(filePath)).toBe(true);

        // Load expected output and compare
        const expectedOutput = loadExpectedOutput('Time_Based_Calendar.ics');
        const actualOutput = fs.readFileSync (filePath, 'utf8');
        // expect(actualOutput).toBe(expectedOutput);
    });

    test('Merge calendar without prefix', async () => {
        const response = await request(server)
            .post('/merge')
            .send({
                linkGroupName: 'No Prefix Calendar',
                calendars: [
                    {
                        url: loadCalendarFile('San_Francisco_Public_Holidays.ics'),
                        prefix: '',
                        override: false,
                    },
                ],
            });

        expect(response.status).toBe(200);
        expect(response.body.url).toMatch(/calendar\/No_Prefix_Calendar/);
        
        // Check if the file was created in the test directory
        const filePath = path.join(TEST_MERGED_CALENDARS_DIR, 'No_Prefix_Calendar.ics');
        expect(fs.existsSync(filePath)).toBe(true);

        // Load expected output and compare
        const expectedOutput = loadExpectedOutput('No_Prefix_Calendar.ics');
        const actualOutput = fs.readFileSync(filePath, 'utf8');
        // expect(actualOutput).toBe(expectedOutput);
    });
    
    test('Merge calendar with override', async () => {
        const response = await request(server)
            .post('/merge')
            .send({
                linkGroupName: 'Override Calendar',
                calendars: [
                    {
                        url: loadCalendarFile('San_Francisco_Public_Holidays.ics'),
                        prefix: 'Override Event',
                        override: true,
                    },
                ],
            });

        expect(response.status).toBe(200);
        expect(response.body.url).toMatch(/calendar\/Override_Calendar/);
        
        // Check if the file was created in the test directory
        const filePath = path.join(TEST_MERGED_CALENDARS_DIR, 'Override_Calendar.ics');
        expect(fs.existsSync(filePath)).toBe(true);

        // Load expected output and compare
        const expectedOutput = loadExpectedOutput('Override_Calendar.ics');
        const actualOutput = fs.readFileSync(filePath, 'utf8');
        // expect(actualOutput).toBe(expectedOutput);
    });

});
