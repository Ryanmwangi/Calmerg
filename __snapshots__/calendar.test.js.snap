// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Merging API Merge calendar with override 1`] = `
{
  "url": "http://127.0.0.1:53391/calendar/Override_Calendar",
}
`;

exports[`Calendar Merging API Merge calendar without prefix 1`] = `
{
  "url": "http://127.0.0.1:53381/calendar/No_Prefix_Calendar",
}
`;

exports[`Calendar Merging API Merge date-based calendar 1`] = `
{
  "url": "http://127.0.0.1:53347/calendar/Date_Based_Calendar",
}
`;

exports[`Calendar Merging API Merge time-based calendar 1`] = `
{
  "url": "http://127.0.0.1:53381/calendar/Time_Based_Calendar",
}
`;
